tools/witx/src/abi.rs

```diff
@@ -836,84 +1602,569 @@ impl<B: Bindgen> Generator<'_, B> {
+        | Type::Builtin(BuiltinType::S32)
+        | Type::Builtin(BuiltinType::U32 { .. })
+        | Type::Builtin(BuiltinType::Char)
+        | Type::Pointer(_)
+        | Type::ConstPointer(_)
+        | Type::Handle(_) => result.push(WasmType::I32),
+
+        Type::Builtin(BuiltinType::U64) | Type::Builtin(BuiltinType::S64) => {
+            result.push(WasmType::I64)
+        }
+        Type::Builtin(BuiltinType::F32) => result.push(WasmType::F32),
+        Type::Builtin(BuiltinType::F64) => result.push(WasmType::F64),
+
+        Type::Record(r) => match r.bitflags_repr() {
+            Some(repr) => result.push(repr.into()),
+            None => {
+                for member in r.members.iter() {
+                    push_wasm(mode, member.tref.type_(), result);
+                }
+            }
+        },
+
+        Type::List(_) => {
+            result.push(WasmType::I32);
+            result.push(WasmType::I32);
+        }
+
+        Type::Buffer(_) => {
+            result.push(WasmType::I32);
+            match mode {
+                CallMode::DefinedExport | CallMode::DeclaredExport => {}
+                CallMode::DefinedImport | CallMode::DeclaredImport => result.push(WasmType::I32),
+            }
+        }
+
+        Type::Variant(v) => {
+            result.push(v.tag_repr.into());
+            let start = result.len();
+            let mut temp = Vec::new();
+
+            // Push each case's type onto a temporary vector, and then merge
+            // that vector into our final list starting at `start`. Note
+            // that this requires some degree of "unification" so we can
+            // handle things like `Result<i32, f32>` where that turns into
+            // `[i32 i32]` where the second `i32` might be the `f32`
+            // bitcasted.
+            for case in v.cases.iter() {
+                let ty = match &case.tref {
+                    Some(ty) => ty,
+                    None => continue,
                 };
-                match r.bitflags_repr() {
-                    Some(IntRepr::U64) => self.emit(&BitflagsFromI64 { ty }),
-                    Some(_) => self.emit(&BitflagsFromI32 { ty }),
-                    None => self.emit(&Load { ty }),
+                push_wasm(mode, ty.type_(), &mut temp);
+
+                for (i, ty) in temp.drain(..).enumerate() {
+                    match result.get_mut(start + i) {
+                        Some(prev) => *prev = unify(*prev, ty),
+                        None => result.push(ty),
+                    }
                 }
             }
-            Type::List(ty) => self.emit(&ListFromPointerLength { ty }),
         }
     }
 }
+
+fn unify(a: WasmType, b: WasmType) -> WasmType {
+    use WasmType::*;
+
+    match (a, b) {
+        (I64, _) | (_, I64) | (I32, F64) | (F64, I32) => I64,
+
+        (I32, I32) | (I32, F32) | (F32, I32) => I32,
+
+        (F32, F32) => F32,
+        (F64, F64) | (F32, F64) | (F64, F32) => F64,
```

Comment:
Passing an f32 value as bitcasted to an f64 type like this looks weird, but thinking through various scenarios, I haven't been able to find anything specifically wrong with it :-}.